File: packages/cli/src/commands/attribution.ts
import { Command } from "commander";
import {
  buildAttributionReport,
  listAttributions,
  loadOrCreateAttribution,
} from "../core/attribution.js";
import { withEpisodeDirectory } from "../core/episode.js";
import { CliError } from "../core/exit.js";

type GroupBy = "predicate" | "taxonomy" | "component";

function parseSinceMs(since?: string): number | null {
  if (!since) return null;
  const match = /^(\d+)([dh])$/.exec(since);
  if (!match) return null;
  const value = Number(match[1]);
  return match[2] === "d"
    ? value * 24 * 60 * 60 * 1000
    : value * 60 * 60 * 1000;
}

function filterSince<T extends { created_at: string }>(
  items: T[],
  since?: string
): T[] {
  const sinceMs = parseSinceMs(since);
  if (!sinceMs) return items;
  const cutoff = Date.now() - sinceMs;
  return items.filter((item) => new Date(item.created_at).getTime() >= cutoff);
}

export function attributionCommand(): Command {
  const attribution = new Command("attribution").description(
    "Explain and report deterministic failure attribution from episodes"
  );

  attribution
    .command("explain")
    .description("Explain failure attribution for one episode")
    .requiredOption("--episode <path>", "Episode directory or bundle")
    .option("--json", "Output JSON instead of text", false)
    .action(async (opts: { episode: string; json?: boolean }) => {
      const result = await withEpisodeDirectory(opts.episode, (episodeDir) =>
        loadOrCreateAttribution(episodeDir)
      );
      if (opts.json) {
        console.log(JSON.stringify(result, null, 2));
      } else {
        console.log("# x-harness Failure Attribution");
        console.log(`- episode_id: ${result.episode_id}`);
        console.log(`- task_id: ${result.task_id}`);
        console.log(`- acceptance_status: ${result.verdict.acceptance_status}`);
        console.log(`- admission_outcome: ${result.verdict.admission_outcome}`);
        if (result.primary) {
          console.log(`- taxonomy: ${result.primary.taxonomy}`);
          console.log(`- predicate: ${result.primary.predicate}`);
          console.log(`- component_id: ${result.primary.component_id}`);
          console.log(`- confidence: ${result.primary.confidence}`);
          console.log(`- rationale: ${result.primary.rationale}`);
        } else {
          console.log("- taxonomy: none");
          console.log("- predicate: none");
        }
      }
    });

  attribution
    .command("report")
    .description("Aggregate failure attribution across episodes")
    .option("--episodes-dir <dir>", "Episodes directory", ".x-harness/episodes")
    .option(
      "--group-by <field>",
      "Group by predicate, taxonomy, or component",
      "predicate"
    )
    .option("--since <duration>", "Filter by age, e.g. 7d or 30d")
    .option("--json", "Output JSON instead of text", false)
    .action(
      async (opts: {
        episodesDir?: string;
        groupBy?: string;
        since?: string;
        json?: boolean;
      }) => {
        if (
          opts.groupBy !== "predicate" &&
          opts.groupBy !== "taxonomy" &&
          opts.groupBy !== "component"
        ) {
          throw new CliError(
            "--group-by must be predicate, taxonomy, or component",
            2
          );
        }
        const attributions = filterSince(
          await listAttributions(opts.episodesDir ?? ".x-harness/episodes"),
          opts.since
        );
        const report = buildAttributionReport(
          attributions,
          opts.groupBy as GroupBy
        );
        if (opts.json) {
          console.log(JSON.stringify(report, null, 2));
        } else {
          console.log("# x-harness Attribution Report");
          console.log(`- group_by: ${report.group_by}`);
          console.log(`- total_episodes: ${report.total_episodes}`);
          console.log(`- withheld_episodes: ${report.withheld_episodes}`);
          console.log(`- unknown_rate: ${report.unknown_rate}`);
          if (report.entropy_warning) {
            console.log(`- entropy_warning: ${report.entropy_warning}`);
          }
          console.log("");
          console.log("## Groups");
          if (report.groups.length === 0) {
            console.log("None.");
          } else {
            for (const group of report.groups) {
              console.log(`- ${group.key}: ${group.count}`);
            }
          }
        }
      }
    );

  return attribution;
}

File: internal/cli/attribution.go
package cli

import (
	"fmt"
	"io"
	"os"

	"github.com/BrianNguyen29/x-harness/internal/attribution"
)

func handleAttribution(args []string, stdout io.Writer, stderr io.Writer) int {
	if len(args) == 0 {
		fmt.Fprintln(stderr, "attribution requires a subcommand: explain, report")
		return ExitUsage
	}

	subcommand := args[0]
	switch subcommand {
	case "explain":
		return handleAttributionExplain(args[1:], stdout, stderr)
	case "report":
		return handleAttributionReport(args[1:], stdout, stderr)
	default:
		fmt.Fprintf(stderr, "unknown attribution subcommand: %s\n", subcommand)
		fmt.Fprintln(stderr, "usage: x-harness attribution explain --episode <dir> [--json]")
		fmt.Fprintln(stderr, "       x-harness attribution report [--episodes-dir <dir>] [--group-by <field>] [--since <duration>] [--json]")
		return ExitUsage
	}
}

func handleAttributionExplain(args []string, stdout io.Writer, stderr io.Writer) int {
	episodeDir := ""
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--episode":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --episode requires a value")
				return ExitUsage
			}
			episodeDir = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if len(args[i]) > 0 && args[i][0] == '-' {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				fmt.Fprintln(stderr, "usage: x-harness attribution explain --episode <dir> [--json]")
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if episodeDir == "" {
		fmt.Fprintln(stderr, "usage: x-harness attribution explain --episode <dir> [--json]")
		return ExitUsage
	}

	info, err := os.Stat(episodeDir)
	if os.IsNotExist(err) {
		fmt.Fprintf(stderr, "error: episode directory does not exist: %s\n", episodeDir)
		return ExitUsage
	}
	if err != nil {
		fmt.Fprintf(stderr, "error: cannot access episode directory: %v\n", err)
		return ExitError
	}
	if !info.IsDir() {
		fmt.Fprintf(stderr, "error: episode path is not a directory: %s\n", episodeDir)
		return ExitUsage
	}

	result, err := attribution.LoadOrCreateAttribution(episodeDir)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	if jsonMode {
		if err := WriteJSON(stdout, result); err != nil {
			return ExitError
		}
		return ExitOK
	}

	fmt.Fprintln(stdout, "# x-harness Failure Attribution")
	fmt.Fprintf(stdout, "- episode_id: %s\n", result.EpisodeID)
	fmt.Fprintf(stdout, "- task_id: %s\n", result.TaskID)
	fmt.Fprintf(stdout, "- acceptance_status: %s\n", result.Verdict.AcceptanceStatus)
	fmt.Fprintf(stdout, "- admission_outcome: %s\n", result.Verdict.AdmissionOutcome)
	if result.Primary != nil {
		fmt.Fprintf(stdout, "- taxonomy: %s\n", result.Primary.Taxonomy)
		fmt.Fprintf(stdout, "- predicate: %s\n", result.Primary.Predicate)
		fmt.Fprintf(stdout, "- component_id: %s\n", result.Primary.ComponentID)
		fmt.Fprintf(stdout, "- confidence: %s\n", result.Primary.Confidence)
		fmt.Fprintf(stdout, "- rationale: %s\n", result.Primary.Rationale)
	} else {
		fmt.Fprintln(stdout, "- taxonomy: none")
		fmt.Fprintln(stdout, "- predicate: none")
	}

	return ExitOK
}

func handleAttributionReport(args []string, stdout io.Writer, stderr io.Writer) int {
	episodesDir := ".x-harness/episodes"
	groupBy := "predicate"
	since := ""
	jsonMode := false

	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--episodes-dir":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --episodes-dir requires a value")
				return ExitUsage
			}
			episodesDir = args[i+1]
			i++
		case "--group-by":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --group-by requires a value")
				return ExitUsage
			}
			groupBy = args[i+1]
			i++
		case "--since":
			if i+1 >= len(args) {
				fmt.Fprintln(stderr, "error: --since requires a value")
				return ExitUsage
			}
			since = args[i+1]
			i++
		case "--json":
			jsonMode = true
		default:
			if len(args[i]) > 0 && args[i][0] == '-' {
				fmt.Fprintf(stderr, "unknown flag: %s\n", args[i])
				fmt.Fprintln(stderr, "usage: x-harness attribution report [--episodes-dir <dir>] [--group-by <field>] [--since <duration>] [--json]")
				return ExitUsage
			}
			fmt.Fprintf(stderr, "unexpected argument: %s\n", args[i])
			return ExitUsage
		}
	}

	if groupBy != "predicate" && groupBy != "taxonomy" && groupBy != "component" {
		fmt.Fprintln(stderr, "--group-by must be predicate, taxonomy, or component")
		return ExitUsage
	}

	attributions, err := attribution.ListAttributions(episodesDir)
	if err != nil {
		fmt.Fprintf(stderr, "error: %v\n", err)
		return ExitError
	}

	attributions = attribution.FilterSince(attributions, since)
	report := attribution.BuildAttributionReport(attributions, groupBy)

	if jsonMode {
		if err := WriteJSON(stdout, report); err != nil {
			return ExitError
		}
		return ExitOK
	}

	fmt.Fprintln(stdout, "# x-harness Attribution Report")
	fmt.Fprintf(stdout, "- group_by: %s\n", report.GroupBy)
	fmt.Fprintf(stdout, "- total_episodes: %d\n", report.TotalEpisodes)
	fmt.Fprintf(stdout, "- withheld_episodes: %d\n", report.WithheldEpisodes)
	fmt.Fprintf(stdout, "- unknown_rate: %g\n", report.UnknownRate)
	if report.EntropyWarning != nil {
		fmt.Fprintf(stdout, "- entropy_warning: %s\n", *report.EntropyWarning)
	}
	fmt.Fprintln(stdout, "")
	fmt.Fprintln(stdout, "## Groups")
	if len(report.Groups) == 0 {
		fmt.Fprintln(stdout, "None.")
	} else {
		for _, group := range report.Groups {
			fmt.Fprintf(stdout, "- %s: %d\n", group.Key, group.Count)
		}
	}

	return ExitOK
}
